### PR TITLE
[expr.sizeof] Clarify padding in class types CWG2609

### LINKED ISSUE #89: CWG2609 [expr.sizeof] The result of applying sizeof to empty class is standard?
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[expr.sizeof] p1 says
> The result of sizeof applied to any of the narrow character types is 1. The result of sizeof applied to **any other fundamental type** ([basic.fundamental]) is **implementation-defined**.

This rule means the result of `sizeof` to a non-narrow-character fundamental type is implementation-defined. The standard says nothing about whether the result of `sizeof` to a class type is implementation-defined or not, it just says 
> When applied to a class, the result is the number of bytes in an object of that class including any padding required for placing objects of that type in an array.

It can only infer that  `sizeof` to a class type that contains a subobject of a non-narrow-character fundamental type is also implementation-defined, as per [expr.sizeof] p1.  This raises a issue, consider that
````cpp
struct S{};
sizeof(S); // all implementations say the result is 1
````
[intro.object] p8 only says
> Unless it is a bit-field, an object with nonzero size shall occupy **one or more bytes** of storage, including every byte that is occupied in full or in part by any of its subobjects.

It didn't specify a concrete number, merely the number is at least one. Such a class type does not contain any subobjects of fundamental types, as aforementioned, the result of `sizeof` to such a type is not clear whether it is guaranteed by standard or specified with implementation-defined. 

All consistent behavior of implementations as if to say the result should be guaranteed by the standard but it was omitted in the standard. 

### Suggested resolution 

we may need to clearly say that the result of `sizeof` applied to any type is **implementation-defined** except for narrow character types.

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes cplusplus/CWG#89

### CODE REVIEW

**Comment by jensmaurer** [comment]
Right.

**Comment by xmh0511** [comment]
Doesn't this clarification negative the consensus in https://github.com/cplusplus/CWG/issues/89#issuecomment-1180032413
> I think what we have here is neither. It's not implementation-defined (no documentation requirement) and it's not unspecified (we shouldn't pretend it might vary from evaluation to evaluation). It's "implementation-defined without documentation requirement", it seems.

**Comment by frederick-vs-ja** [comment]
#5087 suggested that "The representation of a fundamental type is unspecified except as stated in this subclause.". Is such change still suitable?

**Comment by xmh0511** [comment]
> Instead, we say **the size of a class type** is **unspecified**.

The result of `sizeof(S)` where S is a class type is defined as
> When applied to a class, the result is the number of bytes in an object of that class including any padding required for placing objects of that type in an array.

Isn't that just saying the result is the size of the class type? I don't understand what's the difference here.

**Comment by frederick-vs-ja** [comment]
> Either from this perspective or from the normative definitions in [basic.def.odr], they all impose no requirement that all definitions of the same class should have a consistent size unless we can point out the rule.

I think they do strongly imply that the same class **should** have a consistent size. Otherwise, it's even not guaranteed that variable definition like `inline Foo<sizeof(T)> x;` can obey ODR. But it's not clear to me whether there's such a strict requirement.

**Comment by xmh0511** [comment]
> I think they do strongly imply that the same class should have a consistent size. 

AFAIK, the touched rules for your case `inline Foo<sizeof(T)> x;` are [temp.over.link], [basic.def.odr] p14. The former actually cross-references the latter. [basic.def.odr] p14 essentially requires the tokens that form these definitions should be the same, which is regardless of **size**. Also, [basic.def.odr] p14 also requires that the lookup for a name shall refer to the same entity, this is basically regulated by [basic.scope.scope] and [basic.link], in short, these declarations found for the name should correspond and obey [basic.link], again, which is regardless of **size**. The other bullets in [basic.def.odr] p14 also impose no requirement on the size of the entity. 

So, in order to ensure that each evaluation of `sizeof(T)` results in a consistent value, we should explicitly specify it through some other rules instead of relying on the [basic.def.odr] that is hazy/specious for this part.

**Comment by frederick-vs-ja** [comment]
> Also, [basic.def.odr] p14 also requires that the lookup for a name shall refer to the same entity, this is basically regulated by [basic.scope.scope] and [basic.link], in short, these declarations found for the name should correspond and they obey [basic.link], which is regardless of **size** too.

`Foo<sizeof(T)>` refers to different entities if `sizeof(T)` varies.

**Comment by xmh0511** [comment]
> Foo<sizeof(T)> may refer to different entities if sizeof(T) varies.

[temp.over.link]  didn't mention the result of an expression at all, it just says
> Two expressions involving template parameters are considered equivalent **if two function definitions containing the expressions would satisfy the one-definition rule**, except that the tokens used to name the template parameters may differ as long as a token used to name a template parameter in one expression is replaced by another token that names the same template parameter in the other expression.   

`Foo<sizeof(T)>` and `Foo<sizeof(U)>` are equivalent if `T` and `U` are used to name the same template parameter.

**Comment by jensmaurer** [comment]
Obviously, we can only meaningfully talk about the possibly unspecified runtime result of `sizeof` after template instantiation. I have understood `Foo<sizeof(T)>` to mean "any type T that you care to name here", not a reference to a template parameter.

Nobody should be talking about templates here.

**Comment by frederick-vs-ja** [comment]
> Nobody should be talking about templates here.

Perhaps my fault. I meant that `T` is some certain class type.

I should use the aforementioned certain type `S` and `inline char x[sizeof(S)];` in the example, and refer to [basic.link]/(11.1) instead...

**Comment by xmh0511** [comment]
AFAIK, we can only check whether stuff satisfies a particular rule or not rather than that in order to satisfy the rule the stuff must be what it should have been. 
````cpp
char x[size];
````
The rule says `size` must be a constant expression, we just need to check whether `size` violates the rule or not. Instead of assuming that `size` was a constant expression since it appears in a context that requires it to be a constant expression.  

The truth should be that since some rules explicitly say the result of `sizeof(S)` is consistent in a program, the array declared with `sizeof(S)` elements have the same number of elements. The above logic seems to say in order to make these arrays that are declared with `sizeof(S)` have the same number of elements, `sizeof(S)` should always be the same, which is no justification. 

In other words, we can infer the result from conditions but we cannot infer conditions from a result. The result is `5`, so I say the condition must be `1+4`, really?

**Comment by xmh0511** [comment]
> I believe that in order to guarantee that a program containing such declarations well-formed, sizeof(S) cannot vary between instances.

Isn't it that you try to infer the condition from the result? In order to get that result, you predict the condition must be true.

**Comment by frederick-vs-ja** [comment]
> Don't you try to infer the conditions from the result? You mean, in order to get that result, you predict the condition must be true.

My oversight. I meant to say the value should not vary **among all of the aforementioned instances** (that determine the type of `x`), which implies that there are **some** circumstances in which `sizeof(S)` should not change.
(Perhaps it's not good to continue talking about ∃ and ∀ here...)

**Comment by xmh0511** [comment]
> which implies that there are some circumstances in which sizeof(S) should not change.

This should have at least an explicit rule to say the result of `sizeof` applied to a class type is not unspecified. Anyway, in the current modification, the wording says
> The amount and placement of **padding** in a class type is **unspecified**.

Isn't it together with [expr.sizeof] p2
> When applied to a class, the result is the number of **bytes in an object of that class** including any padding required for placing objects of that type in an array.

[basic.types.general] p4
> Bits in the object representation that are not part of the value representation are **padding bits**.

[intro.memory] p1
> A byte is ...  and is composed of a **contiguous sequence of bits**

can conclude that the bytes in an object of class type are unspecified? Again, how many bytes depend on how many bits there are, the number of the latter is unspecified, so the number of the former is as well.

**Comment by jensmaurer** [comment]
Handled by CWG2609
